agent/lib_aws_sdk_php.c

```diff
@@ -14,10 +14,409 @@
+            'BinaryValue' => 'string',
+            'StringListValues' => array('string', ... ),
+            'BinaryListValues' => array('string', ... ),
+            // DataType is required
+            'DataType' => 'string',
+        ),
+        // ... repeated
+    ),
+));
+
+$result = $client->sendMessageBatch(array(
+    // QueueUrl is required
+    'QueueUrl' => 'string',
+    // Entries is required
+    'Entries' => array(
+        array(
+            // Id is required
+            'Id' => 'string',
+            // MessageBody is required
+            'MessageBody' => 'string',
+            'DelaySeconds' => integer,
+            'MessageAttributes' => array(
+                // Associative array of custom 'String' key names
+                'String' => array(
+                    'StringValue' => 'string',
+                    'BinaryValue' => 'string',
+                    'StringListValues' => array('string', ... ),
+                    'BinaryListValues' => array('string', ... ),
+                    // DataType is required
+                    'DataType' => 'string',
+                ),
+                // ... repeated
+            ),
+        ),
+        // ... repeated
+    ),
+));
+
+//clang-format on
+*/
+void nr_lib_aws_sdk_php_sqs_handle(nr_segment_t* segment,
+                                   char* command_name_string,
+                                   size_t command_name_len,
+                                   NR_EXECUTE_PROTO) {
+  char* command_arg_value = NULL;
+  nr_segment_message_params_t message_params = {
+      .library = "SQS",
+      .destination_type = NR_MESSAGE_DESTINATION_TYPE_QUEUE,
+      .messaging_system = "aws_sqs",
+  };
+  nr_segment_cloud_attrs_t cloud_attrs = {0};
+
+  if (NULL == segment) {
+    return;
+  }
+
+  if (NULL == command_name_string || 0 == command_name_len) {
+    return;
+  }
+
+  /* Determine if we instrument this command. */
+  if (command_name_len == AWS_SQS_SEND_MESSAGE_BATCH_COMMAND_LEN
+      && 0
+             == nr_strncmp(AWS_SQS_SEND_MESSAGE_BATCH_COMMAND,
+                           command_name_string,
+                           AWS_SQS_SEND_MESSAGE_BATCH_COMMAND_LEN)) {
+    message_params.message_action = NR_SPANKIND_PRODUCER;
+  } else if (command_name_len == AWS_SQS_SEND_MESSAGE_COMMAND_LEN
+             && 0
+                    == nr_strncmp(AWS_SQS_SEND_MESSAGE_COMMAND,
+                                  command_name_string,
+                                  AWS_SQS_SEND_MESSAGE_COMMAND_LEN)) {
+    message_params.message_action = NR_SPANKIND_PRODUCER;
+  } else if (command_name_len == AWS_SQS_RECEIVE_MESSAGE_COMMAND_LEN
+             && 0
+                    == nr_strncmp(AWS_SQS_RECEIVE_MESSAGE_COMMAND,
+                                  command_name_string,
+                                  AWS_SQS_RECEIVE_MESSAGE_COMMAND_LEN)) {
+    message_params.message_action = NR_SPANKIND_CONSUMER;
+  } else {
+    /* Nothing to do here so exit. */
+    return;
+  }
```

Comment:
This code has so much indirection and requires jumping back and forth to check if the macros have correct values and what those values actually are. Wouldn't this be easier to read and maintain:
```suggestion
#define COMMAND_IS(CMD)                  \
  (command_name_len == (sizeof(CMD) - 1) \
   && 0 == nr_strcmp(CMD, command_name_string))

  /* Determine if we instrument this command. */
  if (COMMAND_IS("sendBatch")) {
    message_params.message_action = NR_SPANKIND_PRODUCER;
  } else if (COMMAND_IS("sendMessage")) {
    message_params.message_action = NR_SPANKIND_PRODUCER;
  } else if (COMMAND_IS("receiveMessage")) {
    message_params.message_action = NR_SPANKIND_CONSUMER;
  } else {
    /* Nothing to do here so exit. */
    return;
  }
#undef IS_COMMAND
```
Suggested change avoids the redirection, and also avoids defining and maintaining the macros for each command and its length.

Reply:
This suggestion is something to consider but is not blocking approval.

Reply:
Agreed, the macro makes the code more readable, but it means we've got to compute the size of a known string every single time we go through the handle_sqs code.

The defines were both to avoid hardcoding values as well as to avoid having to compute sizeof/strlen for those three values every time we go through the code.

For the case where we match the last string, the numbers for computing the size everytime vs having it already computed is:
use macro duration =979650 
precomputed sizes duration =957323

For the case where no strings match:
macro=62778 
precomputed=56180
----------

We could combine approaches to improve readability while maintaining the optimization.

The macro could be something similar to:
```

   #define COMMAND_IS(CMD) \
  (command_name_len == AWS_ ## CMD ## _COMMAND ## _LEN && (nr_streq(AWS_ ## CMD ## _COMMAND, command_name_string)))
```
then the code could look like:
```
  /* Determine if we instrument this command. */
  if (COMMAND_IS(SQS_SEND_MESSAGE_BATCH)) {
    message_params.message_action = NR_SPANKIND_PRODUCER;
  } else if (COMMAND_IS(SQS_SEND_MESSAGE)) {
    message_params.message_action = NR_SPANKIND_PRODUCER;
  } else if (COMMAND_IS(SQS_RECEIVE_MESSAGE)) {
    message_params.message_action = NR_SPANKIND_CONSUMER;
  } else {
    /* Nothing to do here so exit. */
    return;
  }
#undef IS_COMMAND
```

The above improves readability but as seen below this still keeps the performance improvements
macro comparison when len/string matches:
macro that computes lengths each time 924522
macro that uses precomputed lengths 866615

macro comparison when len/string matches:
macro that computes lengths each time 68946
macro that uses precomputed lengths 65715

Reply:
Doesn't the `AWS_ ## CMD ## _COMMAND ## _LEN ` expand to `sizeof (AWS_ ## CMD ## _COMMAND)`? I don't see how this macro:
```
 #define COMMAND_IS(CMD) \
  (command_name_len == AWS_ ## CMD ## _COMMAND ## _LEN && (nr_streq(AWS_ ## CMD ## _COMMAND, command_name_string)))
```
differs from this macro:
```
#define COMMAND_IS(CMD) \
  (command_name_len == (sizeof(CMD) - 1) && nr_streq(CMD, command_name_string))
```
other than it causes preprocessor to expand one more macro at compile time. I don't see how this could have any impact on runtime performance. I really don't see the point in having macro for each command name and for its length if the command string is used in a single place.

Reply:
I ran numerous timing tests on this.

When comparing the condition macro and comparing original format(macro-ized str/len) and when using streq in both cases, the best performer was the original version using name/len precomputed macros.
The conditional macros were pretty close(but the 2nd macro used strncmp).

Maybe it's in the difference of using a macro that is precomputed at runtime and can be used as a constant vs a macro that can't be computed at runtime because it still has unknowns (command_name_len and command_name_string).

The original version using only macro-ized strings and macro-ized lengths is noticeably more performant when using streq.
I've run the benchmarks against this multiple times, and the original code using the len macro is always more performant.

However, when using nrstrcmp (as needed to pass the proper strlength for comparison when using macroized strings), the performance gain is lost.  Given this and the added readability of the macro, after slight modification, I'll put this in and remove the macro-ized strings/lengths.

I'll do a similar thing for class compare.